contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,262 @@
+        emit Slashed(realSlashBnbAmount);
+
+        uint256 _gasLimit = IStakeHub(STAKE_HUB_ADDR).transferGasLimit();
+        (bool success,) = SYSTEM_REWARD_ADDR.call{ gas: _gasLimit, value: realSlashBnbAmount }("");
+        require(success, "TRANSFER_FAILED");
+        return realSlashBnbAmount;
+    }
+
+    /*----------------- view functions -----------------*/
+    /**
+     * @return the amount of shares that corresponds to `_bnbAmount` protocol-controlled BNB.
+     */
+    function getSharesByPooledBNB(uint256 bnbAmount) public view returns (uint256) {
+        if (totalPooledBNB == 0) {
+            return 0;
+        }
+        return (bnbAmount * totalSupply()) / totalPooledBNB;
+    }
+
+    /**
+     * @return the amount of BNB that corresponds to `_sharesAmount` token shares.
+     */
+    function getPooledBNBByShares(uint256 shares) public view returns (uint256) {
+        if (totalSupply() == 0) {
+            return 0;
+        }
+        return (shares * totalPooledBNB) / totalSupply();
+    }
+
+    function unbondRequest(address delegator, uint256 _index) public view returns (UnbondRequest memory, uint256) {
+        bytes32 hash = _unbondRequestsQueue[delegator].at(_index);
+        return (_unbondRequests[hash], _unbondRequestsQueue[delegator].length());
+    }
+
+    function lockedShares(address delegator) public view returns (uint256) {
+        return _lockedShares[delegator];
+    }
+
+    function unbondSequence(address delegator) public view returns (uint256) {
+        return _unbondSequence[delegator].current();
+    }
+
+    function getSelfDelegationBNB() public view returns (uint256) {
+        return getPooledBNBByShares(balanceOf(validator));
+    }
+
+    /*----------------- internal functions -----------------*/
+    function _bootstrapInitialHolder(uint256 initAmount) internal onlyInitializing {
+        assert(validator != address(0));
+        assert(totalSupply() == 0);
+
+        // mint initial tokens to the validator
+        // shares is equal to the amount of BNB staked
+        _mint(validator, initAmount);
+        totalPooledBNB = initAmount;
+        emit Delegated(validator, initAmount, initAmount);
+    }
+
+    function _stake(address delegator, uint256 bnbAmount) internal returns (uint256 shares) {
+        shares = _mintAndSync(delegator, bnbAmount);
+        emit Delegated(delegator, shares, bnbAmount);
+    }
+
+    function _mintAndSync(address account, uint256 bnbAmount) internal returns (uint256 shares) {
+        shares = getSharesByPooledBNB(bnbAmount);
+        _mint(account, shares);
+        totalPooledBNB += bnbAmount;
+    }
+
+    function _burnAndSync(address account, uint256 shares) internal returns (uint256 bnbAmount) {
+        bnbAmount = getPooledBNBByShares(shares);
+        _burn(account, shares);
+        unchecked {
+            // underflow is not possible: totalPooledBNB >= bnbAmount
+            totalPooledBNB -= bnbAmount;
```

Comment:
Let us remove the `unchecked` here,  just for a good sleep.

Reply:
Done